cfme/tests/openstack/infrastructure/test_host_power_control.py

```diff
@@ -19,7 +19,7 @@ def host_collection(appliance):
 @pytest.fixture(scope='module')
 def host_on(host_collection, provider):
     try:
-        my_host_on = host_collection.all(provider)[0]
+        my_host_on = host_collection.all()[0]
```

Comment:
Its not clear here whether or not the collection instance has been filtered by provider already, assuming you actually only want to get hosts for the provider that the fixture is parametrized on. I think it would be good to explicitly check and set the filter within the fixture.

Reply:
Hi MIke,

i am not sure i got your point
there is already a  pytest.mark.provider([OpenstackInfraProvider], scope='module') on the module and for other i just removed "provider" parameter and that was enough

Can you please just point me to the solution because i am not so familiar with all these terms yet.

These regression tests worked before recent changes in infrastructure 

Thanks,
Jad

Reply:
**REQUIRED**
I believe it should look like this:
```python
my_host_on = provider.hosts.all()[0]
```

Reply:
Absolutely.

So, the `pytest.mark.provider` for the module will collect the tests against OpenstackInfraProvider.

There are several pieces at play here, but hopefully I can make this clear.

1. `BaseCollection` type objects (like `host_collection` instance here) have an `all()` method that returns a list of entity instances for that collection
2. By default, this will return all of the hosts - which will be from multiple providers if there are multiple provider configured on the system.
3. `BaseCollection` type objects can be filtered by provider, so that the collection is limited to only those entity instances that have the filtered provider.
4. Before, provider was passed directly to `all` and used to only return entity instances for that provider
5. Now, to get the same behavior, if `all` is implemented to use the filter (it should be!!!) you first have to filter the collection before calling `all`
6. This looks like 
    ```
        host_collection.filter({'provider': provider})
        my_host_on = host_collection.all().pop()
        # the above pop will cause an IndexError if the list is empty, then you can pytest.skip in the except block
    ````

Now, that's the behavior of the collection filters.

@quarckster just responded with a good suggestion to use the hosts collection object that's actually tied to the provider.

This is not necessarily a concern within Jenkins runs, because we default to only allowing 1 provider per appliance, and this test won't be running against an appliance with multiple providers configured and multiple provider hosts listed in the collection.

If you were to run these tests against an appliance with multiple providers configured that all have hosts - picking the first item on an unfiltered collection list *could* give you a host object that's on the wrong provider.